Add support for keywords in searchbar

mounilKshah opened the pull request:
**What does this PR do?**
This PR is to add support for keywords in search bar.

Fixes #638 

Channel discussion:
[Expand search support #T638 #T1258](https://chat.zulip.org/#narrow/stream/206-zulip-terminal/topic/Expand.20search.20support.20.23T638.20.23T1258)

**Tested?**
- [x] Manually
- [x] Existing tests (adapted, if necessary)
- [x] New tests added (for any new behavior)
- [x] Passed linting & tests (each commit)

**Commit flow**
- One commit: Support only for stream and topic keywords

**Notes & Questions**
- This is a WIP PR. It currently only has support for stream and topic name keywords.
- As per the discussions, the plan ahead would be to:
  - Extend support to other keywords for narrows supported by ZT
  - Provide auto-populated text for the current narrow in the search bar
  - Write tests

mounilKshah committed:
model: Add function to extract keywords from search text.

This commit contains Model.extract_keywords_from_search_query()
which is used to extract keywords written in the search query
and sets the narrow accordingly.
Tests are also included in this commit.

neiljp commented on zulipterminal/core.py:
```diff
@@ -492,6 +492,21 @@ def show_media_confirmation_popup(
+                search_query += f" {self.model.narrow[1][0]}:{self.model.narrow[1][1].replace('+', '%2B').replace(' ', '+')}"
+        return search_query
+
     def check_for_prefixes(self, query: str) -> Tuple[str, List[List[str]]]:
```
The name while WIP doesn't matter too much, but now we're firming up the logic, let's make this name more expressive. If I see `check_for_prefixes(something)` in code, it doesn't tell me what it does (I can think of various meanings), or why it returns what it does do - I can look it up in an editor or IDE, but better to read right there :)

So, maybe think what goes in the blank:
```python3
search_query, search_narrow = self.<BLANK>(final_search_query)
```

While some/many ZT functions don't have docstrings (or comments) since they're older, some are very simple and ideally well-named that it may be sufficient to understand what's happening when they're used without them.

You might also rename the other function, if you change the logic of what it does.

neiljp commented on zulipterminal/core.py:
```diff
@@ -501,22 +516,33 @@ def check_for_prefixes(self, query: str) -> Tuple[str, List[List[str]]]:
-                new_narrow.append(word.split(":"))
+                new_narrow.append(
+                    [
+                        word.split(":")[0],
```
We're doing the same word.split operation a lot here, potentially repeatedly, and it likely will appear in other branches I suspect :)

neiljp reviewed (commented):
@mounilKshah This does work better :+1:

I've left feedback mainly on the second commit since it represents better the final state you have right now.

For tests on the current-narrow -> search-text function, the inputs for cases should be achievable from browsing how we handle non-search narrows.

I'd suggest you first split into commits to add each function, then use them together in core.py for 3 commits. Then some tests for each new function in the relevant commit. Covering streams/topics with tests would be good first. More narrow cases may flow quicker once you have the general code/test format :)

mounilKshah committed:
model: Add function to generate additional text for searches.

This commit contains Model.addtional_search_text_for_current_narrow()
which generates string from the current narrow. Tests for the function
have also been added.

neiljp commented on zulipterminal/core.py:
```diff
@@ -492,6 +492,31 @@ def show_media_confirmation_popup(
             self, question, callback, location="center"
         )
 
+    def update_search_text_for_narrow(self, search_query: str) -> str:
```
This looks more complete than the previous version :+1:

Note that all the internal code refers to the model data. This suggests the function belongs in the model.

Also note that we adapt the passed-in string as a return value. It's good to keep each function for one purpose, so leaving the string handling (eg. appending) for the query as something outside of this function.

neiljp commented on tests/core/test_core.py:
```diff
@@ -467,6 +467,50 @@ def test_stream_muting_confirmation_popup(
+            (
+                [["stream", "PTEST PTEST2"]],
+                "FOO",
+                "FOO stream:PTEST+PTEST2",
```
Using inline 'case' ids for the tests would make it clear what you're testing.

neiljp commented on tests/core/test_core.py:
```diff
@@ -467,6 +467,50 @@ def test_stream_muting_confirmation_popup(
+                "FOO is:mentioned",
+            ),
+            (
+                [["pm_with", "some1@email.com", "some2@email.com"]],
```
Is this the correct format? My memory is that each sublist is 2 elements long.

mounilKshah replied:
Usually it is, but in case of `pm_with`, I have always observed the length of the sublist to be 1 + no. of recipients.

neiljp replied:
See https://github.com/zulip/zulip-terminal/blob/main/zulipterminal/model.py#L288, for example.

mounilKshah replied:
Yes, I got it now. I actually didn't realise that all the users under `pm_with` are mentioned under the same string.

neiljp commented on tests/core/test_core.py:
```diff
@@ -467,6 +467,50 @@ def test_stream_muting_confirmation_popup(
+            ),
+        ),
+    )
+    def test_update_search_text_for_narrow(
```
Other narrows: all messages, starred.

neiljp commented on tests/core/test_core.py:
```diff
@@ -467,6 +467,34 @@ def test_stream_muting_confirmation_popup(
+    @pytest.mark.parametrize(
+        "search_text, expected_search_text, expected_search_narrow",
+        [
+            ("FOO", "FOO ", []),
```
I wouldn't expect to see spaces around the output search text.

neiljp commented on tests/core/test_core.py:
```diff
@@ -467,6 +467,34 @@ def test_stream_muting_confirmation_popup(
+            ),
+            ("FOO is:private", "FOO ", [["is", "private"]]),
+            ("FOO is:mentioned", "FOO ", [["is", "mentioned"]]),
+            ("FOO is:starred", "FOO ", [["is", "starred"]]),
```
It'd be great to see some 'space-filled' examples here, like with the other function tests.

neiljp commented on zulipterminal/core.py:
```diff
@@ -492,6 +492,37 @@ def show_media_confirmation_popup(
+                stream_prefix_present_flag = True
+            elif word.startswith("topic:") and stream_prefix_present_flag:
+                new_narrow.append(
+                    ["topic", word.split(":")[1].replace("+", " ").replace("%2B", "+")]
```
It would be good to see various of these replacement operations tested - other than testing the code, it also demonstrates the input->output one might expect, on reading the tests.

neiljp reviewed (commented):
@mounilKshah This looks more readable and complete, including with the tests :+1:

Once you finish the code and tests for the few existing narrows, along with the final (refactor) commit to replace the existing functions with the new functions, this will be mergeable. The functionality may not seem changed, but this refactor will enable supporting more search options in future :)

Note for tests, if you note the changes to tests (separate line) in the git log, there's no need to include it in the commit title.

mounilKshah committed:
core: Update search_messages() for current narrow.

This commit contains function calls used to modify search text
entered, as per the current narrow. Tests have also been updated
accordingly.

neiljp commented on zulipterminal/core.py:
```diff
@@ -531,8 +531,14 @@ def extract_keywords_from_search_query(
+        text_without_keyword, search_narrow = self.extract_keywords_from_search_query(
+            modified_search_text
+        )
+        self.model.narrow = search_narrow
```
My primary remaining concern, given we're potentially taking a pause on this after your work, is this direct setting of the narrow here. Doing so means we potentially don't update other data which would otherwise be set, via eg. model.set_narrow or _narrow_to or something which also updated the UI.

Right now the positive is that the narrow *shouldn't* change, unless a user enters extra search text - and not all combinations work. For now I'd suggest we alert the user with a simple warning popup (popup_with_message), if the search looks like a complicated search which would require setting the narrow, ie. if the old and new narrows wouldn't match.

Examples of cases where this can happen is being in all messages, and adding eg. `is:starred some search text`, in which case the narrow changes. That 'works', but then adding `stream:general` or similar does not. So it'd be good to be consistent for now.

neiljp commented on zulipterminal/model.py:
```diff
@@ -278,6 +278,32 @@ def get_focus_in_current_narrow(self) -> Union[int, Set[None]]:
     def set_focus_in_current_narrow(self, focus_message: int) -> None:
         self.index["pointer"][repr(self.narrow)] = focus_message
 
+    def addtional_search_text_for_current_narrow(self) -> str:
+        """
+        This function is to check if the user is currently inside a narrow or not
+        and if they are, then the additional_search text is created accordingly
```
Can this be simplified/updated?

neiljp commented on zulipterminal/model.py:
```diff
@@ -278,6 +278,32 @@ def get_focus_in_current_narrow(self) -> Union[int, Set[None]]:
+        """
+        # In cases where '+' is already present in stream and/or topic name, it is converted to '%2B'
+        # then the spaces in stream and topic names are converted to '+'
+        additional_search_query_text = ""
```
Minor: The logic here is fairly simple and sets just one primary name, so a much shorter variable name should be fine, within reason - and make it easier to read :) The text gets returned as the name of the function, so the function name can explain what the 'additional_search_query_text` means on it's own.

neiljp commented on tests/model/test_model.py:
```diff
@@ -369,6 +369,47 @@ def test_set_focus_in_current_narrow(self, mocker, model, narrow, msg_id):
+        "narrow, expected_additional_text",
+        (
+            case([], "", id="all_messages"),
+            case([["stream", "PTEST"]], " stream:PTEST", id="stream_without_spaces"),
```
As before, I'd expect the output text to not have any surrounding spaces.

neiljp reviewed (commented):
@mounilKshah The extra commit means I can test this, and it appears to work fairly well from manual testing - and the test cases are reassuring too :+1:

I left some final comments. This is essentially mergeable with some textual adjustments, it's only my concern re unexpectedly safely/unsafely changing the narrow or not, due to this likely being the end of the Summer work on this.

neiljp commented on zulipterminal/model.py:
```diff
@@ -285,6 +285,38 @@ def is_search_narrow(self) -> bool:
         """
         return "search" in [subnarrow[0] for subnarrow in self.narrow]
 
+    def extract_keywords_from_search_query(self, query: str) -> str:
```
This function no longer just extracts keywords, it also sets the base narrow.

neiljp commented on zulipterminal/model.py:
```diff
@@ -278,6 +278,31 @@ def get_focus_in_current_narrow(self) -> Union[int, Set[None]]:
     def set_focus_in_current_narrow(self, focus_message: int) -> None:
         self.index["pointer"][repr(self.narrow)] = focus_message
 
+    def addtional_search_text_for_current_narrow(self) -> str:
```
Spelling?

neiljp commented on tests/core/test_core.py:
```diff
@@ -472,22 +472,28 @@ def test_stream_muting_confirmation_popup(
             (
-                [["pm_with", "foo@zulip.com"], ["search", "BOO"]],
-                [["pm_with", "foo@zulip.com"], ["search", "FOO"]],
+                [["pm_with", "person1@example.com"], ["search", "BOO"]],
+                [["pm_with", "person1@example.com"], ["search", "FOO"]],
             ),
             (
-                [["stream", "PTEST"], ["topic", "RDS"]],
-                [["stream", "PTEST"], ["topic", "RDS"], ["search", "FOO"]],
+                [["stream", "Stream 1"], ["topic", "RDS"]],
+                [["stream", "Stream 1"], ["topic", "RDS"], ["search", "FOO"]],
```
Are these changes just tidying, or required? If tidying, better to keep in a separate commit.

mounilKshah replied:
They are required changes. The `final_narrow` is created only when the narrow exists in the org, else the test case fails.

neiljp reviewed (commented):
@mounilKshah I'd apparently started a review for this but not finished, so just did so. Looks close to a safe refactor.

I thought we might be able to progress further with just search-specific narrow details. We may be able to and you're welcome to investigate, but we may need to refactor the narrow handling more first. Right now that's spread over many locations, possibly with duplicate logic.
